sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala

```diff
@@ -2007,6 +2007,54 @@ class DatasetSuite extends QueryTest
 
     checkAnswer(withUDF, Row(Row(1), null, null) :: Row(Row(1), null, null) :: Nil)
   }
+
+  test("SPARK-34205: Pipe Dataset") {
+    assume(TestUtils.testCommandAvailable("cat"))
+
+    val nums = spark.range(4)
+    val piped = nums.pipe("cat", (l, printFunc) => printFunc(l.toString)).toDF
```

Comment:
@viirya, what do you think about we expose an `transform` equivalent expression exposed as DSL? e.g.)

```scala
scala> val data = Seq((123, "first"), (4567, "second")).toDF("num", "word")
data: org.apache.spark.sql.DataFrame = [num: int, word: string]

scala> data.createOrReplaceTempView("t1")

scala> sql("select transform(*) using 'cat' from t1").show()
+----+------+
| key| value|
+----+------+
| 123| first|
|4567|second|
+----+------+
```

```scala
scala> data.repartition(1).createOrReplaceTempView("t1")

scala> sql("select transform(*) using 'wc -l' as (echo) from t1").show()
+--------+
|    echo|
+--------+
|       2|
+--------+
```

Spark lately added the native support of script transformation, and I think it could do what you want.

Reply:
I think this could address most of comments here such as https://github.com/apache/spark/pull/31296#issuecomment-765877782, being typed or non-standard stuff (as it follows Hive's feature) - at least we have one format to follow, etc.

Reply:
It would be great if we can leverage the script transform #29414.

Reply:
Great point! I don't know how exhaustive Spark implements the Hive's transform feature, but the description in Hive's manual for transform looks pretty much powerful, and much beyond on what we plan to provide with pipe.

https://cwiki.apache.org/confluence/display/Hive/LanguageManual+Transform#LanguageManualTransform-Transform/Map-ReduceSyntax

Looks like the reason of absence of pipe in DataFrame is obvious - transform just replaced it. That looks to be only available in SQL statement so we still need DSL support for using it in SS.

Reply:
I considered `transform` at the beginning as it looks close to pipe. I don't pick it for this because I only see it is exposed as SQL syntax and I am not sure if it works for streaming Dataset? Another reason is that it is designed for untyped Dataset. So if you want to pipe complex object T with custom output instead of column-wise output, "transform"  isn't as powerful as "pipe".

Although I asked our customer and they only use primitive type Dataset for now. So untyped Dataset should be enough for the purpose.

Another reason is although the query looks like "SELECT TRANSFORM(...) FROM ...", it is actually not an expression but implemented as an operator.

Unlike Window function, it seems to me that we cannot have a query like "SELECT a, TRANSFORM(...), c FROM ..." or in DSL format like:

```scala
df.select($"a", $"b", transform(...) ...)
```

But for Window function we can do:

```scala
df.select($"a", $"b", lead("key", 1).over(window) ...)
```

Reply:
> Unlike Window function, it seems to me that we cannot have a query like "SELECT a, TRANSFORM(...), c FROM ..." or in DSL format like:
> 
> ```scala
> df.select($"a", $"b", transform(...) ...)
> ```
> 
> But for Window function we can do:
> 
> ```scala
> df.select($"a", $"b", lead("key", 1).over(window) ...)
> ```
> 
> That being said, in the end it is also `Dataset.transform`, instead of an expression DSL.

I have thought this problem too, first I want to add transform as a DSL function,  in this way, we need to make an equivalent  ScriptTransformation expression first.  We can think that this is just a new expression, or a new function. 


Also add a `Dataset.scriptTransform` could be fine since we can support more flexible usage case and larger scope. Also important, it has a standard to follow and it‘s consistent with SQL.

We don’t necessarily need to write as
```
df.select($"a", $"b", transform(xx))
```
we can write it as 
```
df.scriptTransform(input, script, output...) 
```

Reply:
Personally, I prefer to have ScriptTransformation expression to make this work in the first place. If we find there are more flexible usage needed, then we can add the new API in Dataset. WDYT?

Reply:
Err..I don't think you get the discussed points above. Let me clarify it...

@HyukjinKwon suggested to use "TRANSFORM" for the purpose of piping through external process, instead of adding "pipe" to Dataset API. The idea is basically to add DSL. But the problem is, "TRANSFORM" is not an expression and cannot be used in a DSL approach. So in the end in order to use "TRANSFORM" for piping through external process for streaming Dataset, you will need a top-level API too...But the point of DSL is to avoid a top-level API. So...

Reply:
> Personally, I prefer to have ScriptTransformation expression to make this work in the first place. If we find there are more flexible usage needed, then we can add the new API in Dataset. WDYT?

Can you share how to make transformation as an expression? I don't think It is an expression at all.

Reply:
> So in the end in order to use "TRANSFORM" for piping through external process for streaming Dataset, you will need a top-level API too...But the point of DSL is to avoid a top-level API. So...

Since I did not participate in the community very early, so I don’t know much about the underlying design...
Emmm IMO as a normal developer, we can have things that are useful to users.

>  you will need a top-level API too...

What's the top-level API,  you mean Plan node like `CollectSet` or other thing?

Reply:
I see what @viirya said. I'd agree that transform looks to behave as an operation (not sure that is intended or not, but looks like at least for now) and transform also requires top level API to cover up like we did for `mapPartition`.

If we are OK to add the top level API (again not yet decided so just a 2 cents) then which one? I'd rather say `transform` is something we'd like to be consistent with, instead of `pipe`. They have been exposed as SQL statement, and probably used widely for Spark SQL users, and even Hive users. If we want feature parity then my vote goes to `transform`.

Reply:
```
What's the top-level API, you mean Plan node like CollectSet or other thing?
```
@AngersZhuuuu The top-level API here means the new API added in Dataset.

```
Can you share how to make transformation as an expression? I don't think It is an expression at all.
```
@viirya Sure. I followed the comment "`I have thought this problem too, first I want to add transform as a DSL function, in this way, we need to make an equivalent ScriptTransformation expression first. We can think that this is just a new expression, or a new function`" from @AngersZhuuuu. To add a new expression `ScriptTransformationExpression` for `ScriptTransformation` and turn to `ScriptTransformationExec`.

Two limitations here might need more discussion:
- The script transformation may produce more than one row for a single row, so it cannot use together with other expressions. 
- The script in hive transformation is partition-based, but if we make it an expression, it becomes row based.

Reply:
Thanks @HeartSaVioR. At least I am glad that the discussion can go forward no matter which one you prefer to add.

Honestly I think transform is a weird stuff and it is only for to have pipe feature under SQL syntax. I don't like the transform syntax which is inconvenient to use and verbose. It is not as flexible as pipe's custom print-out function. BTW, for typed dataset, because transform is for untyped, so it is bound to its serialization row format. In the early discussion there are some comments against that, although it is clarified later pipe doesn't suffer from this issue.

If we still cannot get a consensus, maybe I should raise a discussion on dev mailing list to decide pipe or transform top-level API should be added.

@xuanyuanking @AngersZhuuuu The SQL syntax of transform "SELECT TRANSFORM(...)" is pretty confusing. It looks like expression but actually it is an operator, and IMHO you cannot turn it to an expression. If you force it to be an expression, you will create some inconsistency and weird cases. transform is like pipe and their input/output relation is not 1:1 or N:1 but arbitrary.

Reply:
Hmmm, just a clarify, we mean we can add an expression (or function?) like `TRANSFORM`, not convert `TRANSFORM` to it. And we can extract some common logic with `ScriptTransformationExec`.  The usage such as
```
 script_transform(input, script, output)
```
input can be a list of input col such as `a, b, c`
out put can a define such as `col1 string, col2 Int` 
and the return type is  `Array<Struct<col1: String, col2: Int>>` (This DataType can cover all case, and let user to handle)

Then when execute we can make it just run as default format such as `ROW FORMAT DELIMIT`
A simple and general way to implement and then we can add it as a DSL.

Reply:
Do you think this expression can cover all kind of external process output? Transform and pipe have arbitrary relation between input and output. External process can output a line for each input line, and can do aggregation-like output like `wc -l`. I don't know how do you define an expression that the output type is not deterministic.